### PR TITLE
Grab bag of tests

### PR DESCRIPTION
**Pull request by kshyatt**
Added tests for PkgDev error throws in `Pkg`.

Added tests for some missing `isxxx` methods in `iobuffer`.

Added tests for `Irrational`.

### CODE REVIEW

File: test/pkg.jl [hunk 1]
```diff
@@ -182,4 +183,37 @@ temp_pkg_dir() do
         finalize(repo)
         Pkg.update()
     end
+
+    #test PkgDev redirects
+    begin
+        try
+            Pkg.register("IDoNotExist")
```
Comment (tkelman) [inline test/pkg.jl]: these should have a different unexpected error in case this line succeeds somehow

Comment (kshyatt) [inline test/pkg.jl]: ah ok! will fix

Comment (tkelman) [inline test/pkg.jl]: same with the 2 try-catches above that @jakebolewski added for `Pkg.test("IDoNotExist")` and `Pkg.test("IDoNotExist1", "IDoNotExist2")`.

**Comment by tkelman** [comment]
irrational and iobuffer tests should probably be backported, pkgdev tests should not

**Comment by tkelman** [comment]
This is triggering a bug in https://github.com/JuliaLang/julia/blob/9eb74d4730961f52b37ccb8cfdea3b21455e7365/base/rational.jl#L221-L261 on win64, cc @simonbyrne
```
  | | |_| | | | (_| |  |  Version 0.5.0-dev+1476 (2015-11-25 22:03 UTC)
 _/ |\__'_|_|_|\__'_|  |  (HEAD detached at origin/ksh/test)/a6c6f4b* (fork: 3 commits, 0 days)
|__/                   |  x86_64-w64-mingw32

julia> e < big(20 // 6)
false
julia> big(e) < big(20 // 6)
false
julia> x = big(e)
2.718281828459045235360287471352662497757247093699959574966967627724076630353555
julia> y = big(20 // 6)
10//3
julia> isnan(x)
false
julia> isnan(y)
false
julia> xn, xp, xd = Base.decompose(x)
(383787621865837180817076849336057829842706527656748228606440267608139605224109079888716346154450056517601195353758420478403841625929841903940604,-254,1)
julia> yn, yp, yd = Base.decompose(y)
(10,0,3)
julia> xc, yc = widemul(xn,yd), widemul(yn,xd)
(1151362865597511542451230548008173489528119582970244685819320802824418815672327239666149038463350169552803586061275261435211524877789525711821812,10)
julia> xs, ys = sign(xc), sign(yc)
(1,1)
julia> xb, yb = Base.ndigits0z(xc,2) + xp, Base.ndigits0z(yc,2) + yp
(225,4)
```

**Comment by simonbyrne** [comment]
The problem seems to be the first element of the result of [`Base.decompose`](https://github.com/JuliaLang/julia/blob/9eb74d4730961f52b37ccb8cfdea3b21455e7365/base/hashing2.jl#L132-L141).

**Comment by tkelman** [comment]
Hm
```
julia> x
2.718281828459045235360287471352662497757247093699959574966967627724076630353555
julia> s = BigInt()
0
julia> x.prec
256
julia> ccall((:__gmpz_realloc2, :libgmp), Void, (Ptr{BigInt}, Culong), &s, x.prec)
julia> s.size = -fld(-x.prec,(sizeof(Culong)<<3))
8
julia> ccall(:memcpy, Ptr{Void}, (Ptr{Void}, Ptr{Void}, Csize_t), s.d, x.d, s.size*sizeof(Culong))
Ptr{Void} @0x0000000021acb0c0
julia> s, Int(x.exp - x.prec), Int(x.sign)
(199153810457068988487625845466342388628277296048069778809860185741392937567783059768884641227975004982926367923620839421262180583634908156,-254,1)
julia> sizeof(Culong)
4
```

**Comment by tkelman** [comment]
I'm not sure how this hasn't caused any problems sooner, but apparently `GMP_BITS_PER_LIMB` is 64 in our Win64 builds, but `Culong` is 32 bits on Win64 so the `d` fields of `BigInt` and `BigFloat` are different sizes and the `memcpy` isn't right.

**Comment by StefanKarpinski** [comment]
Yikes. I tried to follow through the GMP header files and thought it always ended up being a C unsigned long, but apparently that was incorrect. I guess we need some way of finding GMP's limb size. Might also be a problem for MPFR – I don't recall how crazy that macro definition is.

**Comment by tkelman** [comment]
> some way of finding GMP's limb size

Like https://github.com/JuliaLang/julia/blob/4d0b1ac9c8bc50fe15bdfee29e8fa86ee8809f1e/base/gmp.jl#L23-L37 ?

**Comment by kshyatt** [comment]
@tkelman if I fix the merge conflict and this passes CI, are we ready to go?

**Comment by tkelman** [comment]
Sure. I haven't gotten to writing tests for `decompose` yet (and won't have time to in the near future) but that still needs doing. It's pretty bad that hashing of bigfloats has been wrong on win64 for all of 0.3 because `decompose` has been outputting garbage for as long as it's existed.

**Comment by tkelman** [comment]
needs-tests label as reminder for `Base.decompose`

**Comment by tkelman** [comment]
`Base.decompose` tests added in #15663

**Pull request merged**